File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.cc
#include <algorithm>
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    

    int N;  cin >> N;
    int M;  cin >> M;
    int K;  cin >> K;
    int X;  cin >> X;

    vector<vector<int>> adj(N + 1);

    for(int i = 0 ; i < M ; i++) {
        int s, e;   cin >> s >> e;
        adj[s].push_back(e);
    }


    //BFS

    vector<int> result;
    vector<bool> visited(N + 1, false);

    queue<pair<int, int>> q;
    q.push(make_pair(X, 0));
    visited[X] = true;

    while(!q.empty()) {
        int cur = q.front().first;
        int cnt = q.front().second;
        q.pop();

        if(cnt == K)
            result.push_back(cur);

        for(auto& next : adj[cur]){
            if(!visited[next]){
                visited[next] = true;
                q.push(make_pair(next, cnt + 1));
            }
        }
    }



    if(result.empty())
        cout << -1;
    else{
        sort(result.begin(), result.end());
        for(auto& r : result)
            cout << r << "\n";
    }


}

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class Main {

    static class City{
        int idx;
        int cnt;

        City(int idx, int cnt){
            this.idx = idx;
            this.cnt = cnt;
        }

        int getIdx(){return idx;}
        int getCnt(){return cnt;}
    }
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] NMKX = Arrays.stream(br.readLine().split(" "))
        .mapToInt(Integer::valueOf)
        .toArray();


        Deque<City> dq = new ArrayDeque<>();
        // 간선 정보입력
        List<List<Integer>> adj = new ArrayList<>();
        for(int i = 0 ; i <= NMKX[0] ; i++) adj.add(new ArrayList<>());

        for(int i = 0 ; i < NMKX[1] ; i++){
            int[] fromTo = Arrays.stream(br.readLine().split(" "))
            .mapToInt(Integer::valueOf)
            .toArray();
            adj.get(fromTo[0]).add(fromTo[1]);
        }

        // BFS
        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[NMKX[0] + 1];

        dq.offer(new City(NMKX[3], 0));
        visited[NMKX[3]] = true;

        while(!dq.isEmpty()){
            City cur = dq.pollFirst();

            if(cur.getCnt() > NMKX[2]) break;
            if(cur.getCnt() == NMKX[2]) result.add(cur.getIdx());

            for(int next : adj.get(cur.getIdx())){
                if(!visited[next]){
                    visited[next] = true;
                    dq.offer(new City(next, cur.getCnt() + 1));
                }
            }

        }
        
        if(result.isEmpty()){
            System.out.println(-1);
        }else{
            result.stream()
            .sorted()
            .forEach(System.out::println);
        }


        br.close();
    }
}